### PR TITLE
Add comparison with PyTorch's DataLoader

### PR DESCRIPTION
**Pull request by michael-kuhlmann**
Compares supported features and data throughput between lazy_dataset and torch.utils.data.DataLoader

### CODE REVIEW

**Comment by jensheit** [comment]
should we perhaps add some information on which kind of cpu and gpu the tests were done?

File: comparison/comparison.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+| Example | Feature | `lazy_dataset` | `TorchDataLoader` |
+| ------- | ------- | :------------: | :---------------: |
+|    1    | Dataset class | `lazy_datsaset.Dataset`| `torch_data.Dataset` |
+|    2    | Batching | yes | yes |
+|    2    | Collate batch | manual | automatic |
+|    3    | Shuffle | yes | yes |
+|    4    | Multi-process data loading | yes | yes |
+|    see [README](../README.md)    | Filter (builtins.filter) | yes | no |
+|    5    | Sort by key | yes | no |
```
Comment (boeddeker) [inline comparison/comparison.md]: Can you add `sort by value:  yes no`?

File: comparison/comparison.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+from torch.utils.data import DataLoader as TorchDataLoader
+```
+
+| Example | Feature | `lazy_dataset` | `TorchDataLoader` |
+| ------- | ------- | :------------: | :---------------: |
+|    1    | Dataset class | `lazy_datsaset.Dataset`| `torch_data.Dataset` |
+|    2    | Batching | yes | yes |
+|    2    | Collate batch | manual | automatic |
+|    3    | Shuffle | yes | yes |
+|    4    | Multi-process data loading | yes | yes |
```
Comment (boeddeker) [inline comparison/comparison.md]: Can you add `Multi-thread data loading: yes no`

Comment (boeddeker) [reply comparison/comparison.md]: Maybe it is also good to mention that `Multi-thread` is default

File: comparison/comparison.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+def _prepare_lazy_dataset(dataset, num_workers=0, to_gpu=False):
+    as_lazy_dataset = (
+        dataset
+            .shuffle(rng=np.random.RandomState(0))
+            .batch(16, drop_last=False)
+            .map(Padder(padding_keys=['audio_data']))
+    )
+    if num_workers:
+        as_lazy_dataset = as_lazy_dataset.prefetch(
+            num_workers, 16  # prefetch one batch
```
Comment (boeddeker) [inline comparison/comparison.md]: Can you change this to `num_workers, 2 * num_workers`?

File: comparison/comparison.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    ```
+3. Shuffle
+    ```python
+    >>> data_loader = TorchDataLoader(torch_dict_ds, shuffle=True)
+    >>> dict_ds = dict_ds.shuffle()
+    ```
+4. Parallel data loading
+    ```python
+    >>> data_loader = TorchDataLoader(torch_dict_ds, num_workers=4)
+    >>> dict_ds = dict_ds.prefetch(num_workers=4, buffer_size=4)
```
Comment (boeddeker) [inline comparison/comparison.md]: Can you change this to `dict_ds.prefetch(num_workers=4, buffer_size=8)`

File: comparison/comparison.md (hunk 1 above)
Comment (michael-kuhlmann) [reply comparison/comparison.md]: Considered in commit https://github.com/fgnt/lazy_dataset/pull/27/commits/30012f7a71712eba54a06745121c2b3f399189c5

File: comparison/comparison.md (hunk 2 above)
Comment (michael-kuhlmann) [reply comparison/comparison.md]: Considered in commit https://github.com/fgnt/lazy_dataset/pull/27/commits/30012f7a71712eba54a06745121c2b3f399189c5

File: comparison/comparison.md (hunk 4 above)
Comment (michael-kuhlmann) [reply comparison/comparison.md]: Considered in commit https://github.com/fgnt/lazy_dataset/pull/27/commits/30012f7a71712eba54a06745121c2b3f399189c5

**Comment by michael-kuhlmann** [comment]
> should we perhaps add some information on which kind of cpu and gpu the tests were done?

Added in commit https://github.com/fgnt/lazy_dataset/pull/27/commits/0cfa94218c2aeef324157e9d6172e9f720585142

File: comparison/comparison.md (hunk 3 above)
Comment (michael-kuhlmann) [reply comparison/comparison.md]: Updated in commit https://github.com/fgnt/lazy_dataset/pull/27/commits/2fe45609f1fec46fd23869f478a0648334f37b6d

File: comparison/comparison.md [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+    ```
+2. Batching & Collate
+    ```python
+    >>> data_loader = TorchDataLoader(
+    ...     torch_dict_ds, batch_size=2, collate_fn=None
+    ... )
+    >>> next(iter(data_loader))
+    {'example_id': ['ex_1', 'ex_2'],
+     'observation': [tensor([1, 4]), tensor([2, 5]), tensor([3, 6])],
+     'label': tensor([1, 2])}
```
Comment (boeddeker) [inline comparison/comparison.md]: Can you change the `observation` type to a `numpy.array` for the `torch.utils.data.DataLoader`? This output is strange and using list with different lengths the last element gets dropped:
```
In [25]: examples = { 
    ...:  
    ...:     'ex_1': { 
    ...:         'example_id': 'ex_1', 
    ...:         'observation': [1, 2, 3], 
    ...:         'label': 1 
    ...:     }, 
    ...:     'ex_2': { 
    ...:         'example_id': 'ex_2', 
    ...:         'observation': [5, 6], 
    ...:         'label': 2 
    ...:     } 
    ...:  }
In [26]: torch_dict_ds = TorchDictDataset(examples)                             
In [27]: data_loader = TorchDataLoader(torch_dict_ds, batch_size=2, collate_fn=None)                                                                   
In [28]: next(iter(data_loader))                                                
Out[28]: 
{'example_id': ['ex_1', 'ex_2'],
 'observation': [tensor([1, 5]), tensor([2, 6])],
 'label': tensor([1, 2])}

```

File: comparison/comparison.md [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+
+    >>> dict_ds = dict_ds.batch(2)
+    >>> next(iter(dict_ds))
+    [{'example_id': 'ex_1', 'observation': [1, 2, 3], 'label': 1},
+     {'example_id': 'ex_2', 'observation': [4, 5, 6], 'label': 2}]
+    >>> def collate_fn(batch):
```
Comment (boeddeker) [inline comparison/comparison.md]: Could you replace this function with something like 
```python
    def collate_fn(batch):
        for b in batch[1:]:
            assert batch[0].keys() == b.keys(), batch
        return {k: [b[k] for b in batch] for k in batch[0].keys()}
```
In case someone copies this code, this is more robust.

You could also add a hint, that we have in padertorch (`padertorch.data.utils.collate_fn`) a collate function that can handle nested batches.

File: comparison/comparison.md (hunk 5 above)
Comment (michael-kuhlmann) [reply comparison/comparison.md]: Indeed, that is a weird behavior. I didn't notice it previously. Fixed in https://github.com/fgnt/lazy_dataset/pull/27/commits/33684af06b09d9c28befb42d549b2f7027585261

File: comparison/comparison.md (hunk 6 above)
Comment (michael-kuhlmann) [reply comparison/comparison.md]: Updated in https://github.com/fgnt/lazy_dataset/pull/27/commits/1d9571a7ecb5867874733b2c139555b567826257

**Comment by boeddeker** [comment]
Thanks for the changes. I am now fine with this PR.
